packages/core/src/routes/swagger/utils/general.ts

```diff
@@ -216,19 +220,27 @@ export const validateSwaggerDocument = (document: OpenAPIV3.Document) => {
  * **CAUTION**: This function mutates the input document.
  *
  * Remove operations (path + method) that are tagged with `Cloud only` if the application is not
- * running in the cloud. This will prevent the swagger validation from failing in the OSS
- * environment.
+ * running in the cloud and remove operations with `Dev feature` tag if Logto's `isDevFeatureEnabled` flag
+ * is set to be false.
+ *
+ * This will prevent the swagger validation from failing in the OSS environment.
+ *
  */
-export const removeCloudOnlyOperations = (
+// eslint-disable-next-line complexity
+export const removeUnnecessaryOperations = (
   document: DeepPartial<OpenAPIV3.Document>
 ): DeepPartial<OpenAPIV3.Document> => {
-  if (EnvSet.values.isCloud || !document.paths) {
+  const { isCloud, isDevFeaturesEnabled } = EnvSet.values;
+  if ((isCloud && isDevFeaturesEnabled) || !document.paths) {
     return document;
   }
 
   for (const [path, pathItem] of Object.entries(document.paths)) {
     for (const method of Object.values(OpenAPIV3.HttpMethods)) {
-      if (pathItem?.[method]?.tags?.includes(cloudOnlyTag)) {
+      if (
+        (!isCloud && pathItem?.[method]?.tags?.includes(cloudOnlyTag)) ??
+        (!isDevFeaturesEnabled && pathItem?.[method]?.tags?.includes(devFeatureTag))
+      ) {
```

Comment:
false ?? true = false

Reply:
@simeng-li I think `false ?? true` should still fallback to `true`

Reply:
I think it would be cleaner if we can case them into two Boolean values and use the `OR` operator `||`

Reply:
I believe I was using `||` and the linter used `??` to replace this.

Reply:
Updated